homeassistant/util/async_.py

```diff
@@ -76,3 +74,36 @@ def run_callback() -> None:
 
     loop.call_soon_threadsafe(run_callback)
     return future
+
+
+async def safe_wait(
+    tasks: Any, logger: Optional[logging.Logger] = None, return_exceptions: bool = False
+) -> List[Any]:
+    """Safe version of wait and gather.
+
+    It work like gather but wait don't break any workflows.
+    It allow also to log exception in correct namespace.
+    """
+    all_tasks = list(tasks)
+    if not all_tasks:
+        return []
+    finished_tasks, _ = await asyncio.wait(all_tasks)
+
+    results: List[Any] = []
+    raise_exception: Optional[BaseException] = None
+    for task in finished_tasks:
+        if not task.done():
```

Comment:
Given we're invoking `wait` with the `return_when=ALL_COMPLETED` default, when would this ever happen?  Is there a unit test that invokes this fork of the code?  Additionally, aren't incomplete tasks in the other element of the tuple (captured in the `_` variable) and wouldn't be in the `finished_tasks` anyway?

Reply:
that is not 100% correct. A task can be cancel by others. That is a real top-level implementation what need respect any kind of issues

Reply:
So take a look at the [docs for `done()`](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.done): 
>> A Task is done when the wrapped coroutine either returned a value, raised an exception, or the Task was cancelled.

I think this code is not going to work as expected.  Unit tests please. :)

Reply:
I see, wrong function. Fixed